File: android/src/main/java/com/lufinkey/react/eventemitter/RNModuleEvents.java
package com.lufinkey.react.eventemitter;

import com.facebook.react.bridge.Callback;

import java.util.ArrayList;
import java.util.HashMap;

public class RNModuleEvents {
	HashMap<String, ArrayList<RNEventCallback>> eventListeners = new HashMap<>();

	public void addListener(String eventName, Callback callback, boolean once) {
		synchronized (eventListeners) {
			ArrayList<RNEventCallback> listeners = eventListeners.get(eventName);
			if (listeners == null) {
				listeners = new ArrayList<>();
				eventListeners.put(eventName, listeners);
			}
			listeners.add(new RNEventCallback(callback, once));
		}
	}

	public void prependListener(String eventName, Callback callback, boolean once) {
		synchronized (eventListeners) {
			ArrayList<RNEventCallback> listeners = eventListeners.get(eventName);
			if (listeners == null) {
				listeners = new ArrayList<>();
				eventListeners.put(eventName, listeners);
			}
			listeners.add(0, new RNEventCallback(callback, once));
		}
	}

	public void removeListener(String eventName, Callback callback) {
		synchronized (eventListeners) {
			ArrayList<RNEventCallback> listeners = eventListeners.get(eventName);
			for (int i=0; i<listeners.size(); i++) {
				RNEventCallback listener = listeners.get(i);
				if (listener.getCallback() == callback) {
					listeners.remove(i);
					return;
				}
			}
		}
	}

	public void removeAllListeners(String eventName) {
		synchronized (eventListeners) {
			if (eventName == null) {
				eventListeners.clear();
			}
			else {
				eventListeners.remove(eventName);
			}
		}
	}

	public int getListenerCount(String eventName) {
		synchronized (eventListeners) {
			ArrayList<RNEventCallback> listeners = eventListeners.get(eventName);
			if (listeners == null) {
				return 0;
			}
			return listeners.size();
		}
	}

	public boolean emit(String eventName, Object... args) {
		ArrayList<RNEventCallback> tmpListeners = null;

		synchronized (eventListeners) {
			ArrayList<RNEventCallback> listeners = eventListeners.get(eventName);
			if(listeners != null) {
				tmpListeners = new ArrayList<>(listeners);
			}

			// remove "once" event listeners
			if(listeners != null) {
				for (int i = 0; i < listeners.size(); i++) {
					RNEventCallback listener = listeners.get(i);
					if (listener.isCalledOnlyOnce()) {
						listeners.remove(i);
						i--;
					}
				}
			}
		}

		if (tmpListeners != null && tmpListeners.size() > 0) {
			//invoke events
			for (int i = 0; i < tmpListeners.size(); i++) {
				RNEventCallback listener = tmpListeners.get(i);
				listener.getCallback().invoke(args);
			}
			return true;
		}
		return false;
	}
}

File: ios/RNModuleEvents.m

#import "RNModuleEvents.h"
#import "RNEventCallback.h"

@interface RNModuleEvents() {
	NSMutableDictionary<NSString*, NSMutableArray<RNEventCallback*>*>* _eventListeners;
}
@end


@implementation RNModuleEvents

-(id)init {
	if(self = [super init]) {
		_eventListeners = [NSMutableDictionary dictionary];
	}
	return self;
}

-(void)addListener:(void(^)(NSArray*))listener forEvent:(NSString*)event onlyOnce:(BOOL)once {
	@synchronized (_eventListeners) {
		NSMutableArray<RNEventCallback*>* listeners = _eventListeners[event];
		if (listeners == nil) {
			listeners = [[NSMutableArray alloc] init];
			_eventListeners[event] = listeners;
		}
		[listeners addObject:[[RNEventCallback alloc] initWithBlock:listener callOnlyOnce:once]];
	}
}

-(void)prependListener:(void(^)(NSArray*))listener forEvent:(NSString*)event onlyOnce:(BOOL)once {
	@synchronized (_eventListeners) {
		NSMutableArray<RNEventCallback*>* listeners = _eventListeners[event];
		if (listeners == nil) {
			listeners = [[NSMutableArray alloc] init];
			_eventListeners[event] = listeners;
		}
		[listeners insertObject:[[RNEventCallback alloc] initWithBlock:listener callOnlyOnce:once] atIndex:0];
	}
}

-(void)removeListener:(void(^)(NSArray*))listener forEvent:(NSString*)event {
	@synchronized (_eventListeners) {
		NSMutableArray<RNEventCallback*>* listeners = _eventListeners[event];
		for (NSUInteger i=0; i<listeners.count; i++) {
			RNEventCallback* callback = listeners[i];
			if (callback.block == listener) {
				[listeners removeObjectAtIndex:i];
				return;
			}
		}
	}
}

-(void)removeAllListenersForEvent:(NSString*)event {
	@synchronized (_eventListeners) {
		if(event == nil) {
			[_eventListeners removeAllObjects];
		}
		else {
			[_eventListeners removeObjectForKey:event];
		}
	}
}

-(void)removeAllListeners {
	@synchronized (_eventListeners) {
		[_eventListeners removeAllObjects];
	}
}

-(NSUInteger)listenerCountForEvent:(NSString*)event {
	@synchronized (_eventListeners) {
		NSMutableArray<RNEventCallback*>* listeners = _eventListeners[event];
		if (listeners == nil) {
			return 0;
		}
		return listeners.count;
	}
}

-(BOOL)emitEvent:(NSString*)event withParams:(NSArray*)params {
	NSArray<RNEventCallback*>* tmpListeners = nil;
	
	@synchronized (_eventListeners) {
		NSMutableArray<RNEventCallback*>* listeners = _eventListeners[event];
		if(listeners != nil) {
			tmpListeners = [NSMutableArray arrayWithArray:listeners];
		}
		
		// remove "once" event listeners
		if(listeners != nil) {
			for (NSUInteger i=0; i<listeners.count; i++) {
				RNEventCallback* listener = listeners[i];
				if (listener.calledOnlyOnce) {
					[listeners removeObjectAtIndex:i];
					i--;
				}
			}
		}
	}
	
	if (tmpListeners != nil && tmpListeners.count > 0) {
		//invoke events
		for(NSUInteger i=0; i<tmpListeners.count; i++) {
			RNEventCallback* listener = tmpListeners[i];
			listener.block(params);
		}
		return true;
	}
	return false;
}

@end
